flink-connectors/flink-connector-kafka-base/src/test/java/org/apache/flink/streaming/connectors/kafka/KafkaMigrationTestBase.java

```diff
@@ -0,0 +1,171 @@
+		new TypeInformationSerializationSchema<>(BasicTypeInfo.INT_TYPE_INFO, new ExecutionConfig());
+	protected final KeyedSerializationSchema<Integer> integerKeyedSerializationSchema =
+		new KeyedSerializationSchemaWrapper<>(integerSerializationSchema);
+
+	/**
+	 * TODO change this to the corresponding savepoint version to be written (e.g. {@link MigrationVersion#v1_3} for 1.3)
+	 * TODO and remove all @Ignore annotations on write*Snapshot() methods to generate savepoints
+	 * TODO Note: You should generate the savepoint based on the release branch instead of the master.
+	 */
+	protected final Optional<MigrationVersion> flinkGenerateSavepointVersion = Optional.empty();
+
+	public KafkaMigrationTestBase(MigrationVersion testMigrateVersion) {
+		this.testMigrateVersion = checkNotNull(testMigrateVersion);
+	}
+
+	public String getOperatorSnapshotPath() {
+		return getOperatorSnapshotPath(testMigrateVersion);
+	}
+
+	public String getOperatorSnapshotPath(MigrationVersion version) {
+		return "src/test/resources/kafka-migration-kafka-producer-flink-" + version + "-snapshot";
+	}
+
+	/**
+	 * Override {@link KafkaTestBase}. Kafka Migration Tests are starting up Kafka/ZooKeeper cluster manually
+	 */
+	@BeforeClass
+	public static void prepare() throws Exception {
+	}
+
+	/**
+	 * Override {@link KafkaTestBase}. Kafka Migration Tests are starting up Kafka/ZooKeeper cluster manually
+	 */
+	@AfterClass
+	public static void shutDownServices() throws Exception {
+	}
+
+	/**
+	 * Manually run this to write binary snapshot data.
+	 */
+	@Ignore
+	@Test
+	public void writeSnapshot() throws Exception {
+		try {
+			checkState(flinkGenerateSavepointVersion.isPresent());
+			startClusters();
+
+			OperatorSubtaskState snapshot = initializeTestState();
+			OperatorSnapshotUtil.writeStateHandle(snapshot, getOperatorSnapshotPath(flinkGenerateSavepointVersion.get()));
+		}
+		finally {
+			shutdownClusters();
+		}
+	}
+
+	private OperatorSubtaskState initializeTestState() throws Exception {
+		try (OneInputStreamOperatorTestHarness testHarness = createTestHarness()) {
+			testHarness.setup();
+			testHarness.open();
+
+			// Create a committed transaction
+			testHarness.processElement(42, 0L);
+
+			// TODO: when stop with savepoint is available, replace this code with it (with stop with savepoint
+			// there won't be any pending transactions)
+			OperatorSubtaskState snapshot = testHarness.snapshot(0L, 1L);
+			// We kind of simulate stop with savepoint by making sure that notifyOfCompletedCheckpoint is called
+			testHarness.notifyOfCompletedCheckpoint(0L);
+
+			// Create a Pending transaction
+			testHarness.processElement(43, 2L);
+			return snapshot;
+		}
+	}
+
+	@SuppressWarnings("warning")
+	@Test
+	public void testRestoreProducer() throws Exception {
+		try {
+			startClusters();
+
+			initializeTestState();
+
+			try (OneInputStreamOperatorTestHarness testHarness = createTestHarness()) {
```

Comment:
Nit pick: Can we use a different transaction id for the restored producer?

It seems this test is a bit tricky.

FlinkKafkaProducer may abort the transactions in two different cases. (1) when there is no checkpoint, FlinkKafkaProducer aborts all the transactions of a large range based on `safeScaleDownFactor`. (2) when there are pending transactions in the checkpoint. Case (2) is what we expect here.

Because the snapshot is a little difficult to inspect, in order to rule out case (1), should we use a different transactional id when creating the restored producer?

Reply:
Good question. I'm not sure if we can change this behaviour with altering production code. However...

... regardless if (1) or (2) has happened here, from the perspective of this use case/scenario, the result is correct in both cases. However there is a different test that focus on (1) and (2) and differentiates between them `FlinkKafkaProducer011ITCase#testFailBeforeNotifyAndResumeWorkAfterwards`. 

I guess for completeness it would be best to add more test cases here, but it would prolong already very length tests and it depends what are we planning to support? So far in the long term we were aiming to "officially" support kafka migration stories only with "stop with savepoint" feature (in progress). Ideally we would want "stop with savepoint without side-effects" (no ongoing transactions) but work on this hasn't yet started.

Reply:
Yeah, I think the test is OK. It just took me some time to make sure we are testing something we actually want to test. The patch LGTM.